sdk/dotnet/Pulumi/Testing/Mocks.cs

```diff
@@ -0,0 +1,210 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Google.Protobuf.Collections;
+using Google.Protobuf.WellKnownTypes;
+using Pulumi.Serialization;
+using Pulumirpc;
+
+namespace Pulumi.Testing
+{
+    /// <summary>
+    /// Hooks to mock the engine and provide test doubles for offline unit testing of stacks.
+    /// </summary>
+    public interface IMocks
+    {
+        /// <summary>
+        /// Invoked when a new resource is created by the program.
+        /// </summary>
+        /// <param name="resource">Resource instance.</param>
+        /// <param name="type">Resource type name.</param>
+        /// <param name="name">Resource name.</param>
+        /// <param name="inputs">Dictionary of resource input properties.</param>
+        /// <param name="provider">Provider.</param>
+        /// <param name="id">Resource identifier.</param>
+        /// <returns>A tuple of a resource identifier and resource state. State can be either a POCO
+        /// or a dictionary bag.</returns>
+        Task<(string id, object state)> NewResourceAsync(Resource resource, string type, string name,
+            ImmutableDictionary<string, object> inputs, string? provider, string? id);
+
+        /// <summary>
+        /// Invoked when the program needs to call a provider to load data (e.g., to retrieve an existing
+        /// resource).
+        /// </summary>
+        /// <param name="token">Function token.</param>
+        /// <param name="args">Dictionary of input arguments.</param>
+        /// <param name="provider">Provider.</param>
+        /// <returns>Invocation result, can be either a POCO or a dictionary bag.</returns>
+        Task<object> CallAsync(string token, ImmutableDictionary<string, object> args, string? provider);
+
+        /// <summary>
+        /// Invoked when an error is raised by the program execution.
+        /// </summary>
+        /// <param name="message">Error message.</param>
+        Task HandleErrorAsync(string message);
+
+        /// <summary>
+        /// Invoked when a resource registers its outputs.
+        /// </summary>
+        /// <param name="urn">Resource URN.</param>
+        /// <param name="outputs">Outputs property bag.</param>
+        Task ResourceOutputsAsync(string urn, ImmutableDictionary<string, object> outputs);
```

Comment:
I don't think that this is necessary--this is essentially an implementation detail of the engine.

Reply:
The point is to get the actual outputs of a stack in https://github.com/pulumi/pulumi/pull/3696/files#diff-c87e9a22bf55d4ae8e2b14e128465573R39
I agree this may be a bit too low-level for the user code. We can hide it from the public API surface.

Reply:
Removed